File: app/src/test/java/com/matchthree/game/Boards.kt
package com.matchthree.game

import com.matchthree.game.model.Board
import com.matchthree.game.model.Gem
import com.matchthree.game.model.GemType
import com.matchthree.game.model.Position

/**
 * Builds a [Board] from rows of single-character codes for readable test fixtures:
 * R=RED, G=GREEN, B=BLUE, Y=YELLOW, A/other=PURPLE or ORANGE via [charToType],
 * '.' = empty cell. Gem ids are assigned row-major starting at [idOffset].
 */
object Boards {

    fun fromRows(vararg rows: String): Board = fromRows(rows.toList())

    fun fromRows(rows: List<String>, idOffset: Int = 0): Board {
        require(rows.isNotEmpty()) { "at least one row" }
        val width = rows[0].length
        require(rows.all { it.length == width }) { "all rows must have equal width" }
        var nextId = idOffset
        return Board.create(width, rows.size) { pos: Position ->
            charToType(rows[pos.row][pos.col])?.let { Gem(nextId++, it) }
        }
    }

    fun charToType(char: Char): GemType? = when (char) {
        'R' -> GemType.RED
        'G' -> GemType.GREEN
        'B' -> GemType.BLUE
        'Y' -> GemType.YELLOW
        'P', 'A' -> GemType.PURPLE
        'O' -> GemType.ORANGE
        '.' -> null
        else -> error("unknown gem char: $char")
    }

    /** Reverse of [charToType]; PURPLE renders as 'P'. */
    fun typeToChar(type: GemType): Char = when (type) {
        GemType.RED -> 'R'
        GemType.GREEN -> 'G'
        GemType.BLUE -> 'B'
        GemType.YELLOW -> 'Y'
        GemType.PURPLE -> 'P'
        GemType.ORANGE -> 'O'
    }
}

File: godot/Tests/Boards.cs
using MatchThree.Engine.Model;

namespace MatchThree.Engine.Tests;

/// <summary>
/// Builds a <see cref="Board"/> from rows of single-character codes for readable
/// test fixtures: R=RED, G=GREEN, B=BLUE, Y=YELLOW, A/other=PURPLE or ORANGE via
/// <see cref="CharToType"/>, '.' = empty cell. Gem ids are assigned row-major
/// starting at <paramref name="idOffset"/>.
/// </summary>
public static class Boards
{
    public static Board FromRows(params string[] rows) => FromRows(rows.ToList());

    public static Board FromRows(IEnumerable<string> rows, int idOffset = 0)
    {
        var rowList = rows.ToList();
        if (rowList.Count == 0)
            throw new ArgumentException("at least one row", nameof(rows));
        var width = rowList[0].Length;
        if (rowList.Any(r => r.Length != width))
            throw new ArgumentException("all rows must have equal width", nameof(rows));

        var nextId = idOffset;
        return Board.Create(width, rowList.Count, pos =>
        {
            var type = CharToType(rowList[pos.Row][pos.Col]);
            return type is { } t ? new Gem(nextId++, t) : (Gem?)null;
        });
    }

    public static GemType? CharToType(char c) => c switch
    {
        'R' => GemType.Red,
        'G' => GemType.Green,
        'B' => GemType.Blue,
        'Y' => GemType.Yellow,
        'P' or 'A' => GemType.Purple,
        'O' => GemType.Orange,
        '.' => null,
        _ => throw new ArgumentException($"unknown gem char: {c}"),
    };

    /// <summary>Reverse of <see cref="CharToType"/>; PURPLE renders as 'P'.</summary>
    public static char TypeToChar(GemType type) => type switch
    {
        GemType.Red => 'R',
        GemType.Green => 'G',
        GemType.Blue => 'B',
        GemType.Yellow => 'Y',
        GemType.Purple => 'P',
        GemType.Orange => 'O',
        _ => throw new ArgumentOutOfRangeException(nameof(type)),
    };
}